Cancel before substituting in Pow._eval_nseries

0sidharth opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #21075 

#### Brief description of what is fixed or changed
The original expression is now expanded and compared with the resultant expression obtain in the series expansion. Previously, values were substituted for zero checking on the difference of the expressions, which lead to corner cases and bugs.
#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * Fixed a bug in `Pow._eval_nseries` that added `Order` terms to exact expansions
<!-- END RELEASE NOTES -->

jksuom commented on sympy/core/power.py:
```diff
@@ -1694,7 +1694,7 @@ def mul(d1, d2):
 
         for i in (1, 2, 3):
-            if (res - self).subs(x, i) is not S.Zero:
+            if (res - self).cancel().subs(x, i) is not S.Zero:
```
There may be other possibilities, e.g `expand` or `_mexpand`. We should choose one that is most efficient in general. (Many tests may be needed though.)

jksuom replied:
This is supposed to be a quick test.

jksuom replied:
An order term should be added whenever the expansion of `base**exp` is truncated. That will not happen when `exp` is a positive integer and sufficiently small in comparison with `n`, In that case, it seems that `_mexpand` is exactly what is needed to transform `base**exp`. Hence I would consider something like this (instead of the `subs` trick):
```
        if not (exp.is_integer and exp.is_positive and exp*d < n and
                res == _mexpand(self)):
            res += O(x**n, x)
```
(The purpose of `exp*d < n` is to avoid expanding something like `base**10000`.)

0sidharth replied:
Would this check take care of the `base**10000` case?https://github.com/sympy/sympy/blob/62572863590b1c3363d6268f75aea6e83e385728/sympy/core/power.py#L1623-L1625

jksuom replied:
It is possible. Maybe you could run tests.

jksuom replied:
However, `m` may be 0 or negative.

0sidharth replied:
The maxpow condition was indeed insufficient. An example is https://github.com/sympy/sympy/blob/62572863590b1c3363d6268f75aea6e83e385728/sympy/series/tests/test_limits.py#L725-L726 After this change, this example runs almost instantaneously compared to taking a couple of seconds earlier

0sidharth commented on sympy/core/power.py:
```diff
@@ -1622,7 +1622,7 @@ def _eval_nseries(self, x, n, logx, cdir=0):
         maxpow = n - m*e
 
-        if maxpow < S.Zero:
+        if maxpow.is_negative:
```
This change is with ref to: https://github.com/sympy/sympy/issues/20897

jksuom replied:
Did you encounter an example where this would raise?

0sidharth replied:
I had encountered the case with the condition on `Mul._eval_nseries` changed, but not with the current condition. I think it should be reproducible, will try and get back to you.

jksuom replied:
It looks like `n`, `m` and `e` would be real numbers. (Otherwise there is probably a bug somewhere in the leading term code.)

0sidharth replied:
They are indeed real, but zero checking isn't always possible as Oscar had pointed out in the issue.
```py
In [1]: a = (5*log(5)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) + 1)

In [2]: a.evalf()
Out[2]: 0.e-125

In [3]: if a < S.Zero:
   ...:     print(".")
   ...: 
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-b716c5be69d4> in <module>
----> 1 if a < S.Zero:
      2     print(".")

~/sympy/sympy/core/relational.py in __bool__(self)
    396 
    397     def __bool__(self):
--> 398         raise TypeError("cannot determine truth value of Relational")
    399 
    400     def _eval_as_set(self):

TypeError: cannot determine truth value of Relational
```

0sidharth replied:
While not the simplest form, this example raises an error on the current master -
```py
In [1]: w, p = symbols('w p')

In [2]: logx = -p*log(x) + 6*p*log(3*p + 1) - 5*p*log(5*p + 2) - 6*p*log(6) + 6*p*log(2)

In [3]: x = w

In [4]: a = 1/(180*p**2*w**(5*log(5)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)))*sqrt(3*p + 1)*exp(5*p*log(5)*log(p)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*lo
   ...: g(3)) - 30*p*log(5)*log(3*p + 1)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) + 25*p*log(5)*log(5*p + 2)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3))
   ...: + 30*p*log(5)*log(6)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) - 30*p*log(2)*log(5)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) + 5*p*log(5)) + 12
   ...: 0*w**(5*log(5)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)))*p*sqrt(3*p + 1)*exp(5*p*log(5)*log(p)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) - 30*p
   ...: *log(5)*log(3*p + 1)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) + 25*p*log(5)*log(5*p + 2)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) + 30*p*log(5
   ...: )*log(6)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) - 30*p*log(2)*log(5)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) + 5*p*log(5)) + 20*w**(5*log(5
   ...: )/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)))*sqrt(3*p + 1)*exp(5*p*log(5)*log(p)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) - 30*p*log(5)*log(3*p
   ...:  + 1)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) + 25*p*log(5)*log(5*p + 2)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) + 30*p*log(5)*log(6)/(-6*lo
   ...: g(6) - 5*log(5) + 6*log(2) + 6*log(3)) - 30*p*log(2)*log(5)/(-6*log(6) - 5*log(5) + 6*log(2) + 6*log(3)) + 5*p*log(5)))

In [5]: a.nseries(x, 1, logx)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-eb9556bec087> in <module>
----> 1 a.nseries(x, 1, logx)

~/sympy/sympy/core/expr.py in nseries(self, x, x0, n, dir, logx, cdir)
   3335             return self
   3336         if x is None or x0 or dir != '+':  # {see XPOS above} or (x.is_positive == x.is_negative == None):
-> 3337             return self.series(x, x0, n, dir, cdir=cdir)
   3338         else:
   3339             return self._eval_nseries(x, n=n, logx=logx, cdir=cdir)

~/sympy/sympy/core/expr.py in series(self, x, x0, n, dir, logx, cdir)
   2971                 rep2 = x
   2972                 rep2b = -x0
-> 2973             s = self.subs(x, rep).series(x, x0=0, n=n, dir='+', logx=logx, cdir=cdir)
   2974             if n is None:  # lseries...
   2975                 return (si.subs(x, rep2 + rep2b) for si in s)

~/sympy/sympy/core/expr.py in series(self, x, x0, n, dir, logx, cdir)
   2981             # replace x with an x that has a positive assumption
   2982             xpos = Dummy('x', positive=True, finite=True)
-> 2983             rv = self.subs(x, xpos).series(xpos, x0, n, dir, logx=logx, cdir=cdir)
   2984             if n is None:
   2985                 return (s.subs(xpos, x) for s in rv)

~/sympy/sympy/core/expr.py in series(self, x, x0, n, dir, logx, cdir)
   2988 
   2989         if n is not None:  # nseries handling
-> 2990             s1 = self._eval_nseries(x, n=n, logx=logx, cdir=cdir)
   2991             o = s1.getO() or S.Zero
   2992             if o:

~/sympy/sympy/core/power.py in _eval_nseries(self, x, n, logx, cdir)
   1623         maxpow = n - m*e
   1624 
-> 1625         if maxpow < S.Zero:
   1626             return O(x**(m*e), x)
   1627 

~/sympy/sympy/core/relational.py in __bool__(self)
    396 
    397     def __bool__(self):
--> 398         raise TypeError("cannot determine truth value of Relational")
    399 
    400     def _eval_as_set(self):

TypeError: cannot determine truth value of Relational

```
I will try to reproduce the issue with a simpler example and add it as a test.

0sidharth replied:
If there are no further suggested changes in this PR, think I will revert the relational operators -> assumptions changes and look into this in more detail, and open a new PR/close the issue according to what conclusion is drawn, because it is also unrelated to this PR.

0sidharth committed:
Fixed a bug in Pow._eval_nseries

Check if expanded form of original expression is same as obtained expression
instead of zero checking by subtituting values into the difference

0sidharth committed:
Pow nseries: replace relational operators with assumptions

jksuom commented:
Looks good. Thanks!
